### PR TITLE
Improve dedupe script

### LINKED ISSUE #3686: [BUG] binary-dedupe doesn't fail the build on errors
**Describe the bug**
binary-dedupe uses ant exec task. failonerror is false by defaut

**Steps/Code to reproduce bug**
introduce any sort of error situation, or manually inject an non-zero `exit 123`, the build proceeds

**Expected behavior**
errors in binary-dedupe.sh, especially validation of bitwise-identity must fail the build

**Environment details (please complete the following information)**
- local dev

### PR DESCRIPTION
**Pull request by gerashegalov**
closes #3686 

- bug: fail build on error is missing
- make it skippable via environment variable `SKIP_BINARY_DEDUPE`
- replace md5 with sha1 https://www.avira.com/en/blog/md5-the-broken-algorithm
- replace diff with sha1 
- more general dedupe logic leaving only a single copy of files with a single hash value 

Signed-off-by: Gera Shegalov <gera@apache.org>

### CODE REVIEW

**Comment by gerashegalov** [comment]
build

**Comment by gerashegalov** [comment]
pre-merge errored out in a seemingly unrelated place:
```
[2021-09-28T16:23:20.060Z] + python /home/jenkins/agent/workspace/jenkins-rapids_premerge-github-2855-ci-2/integration_tests/runtests.py --rootdir /home/jenkins/agent/workspace/jenkins-rapids_premerge-github-2855-ci-2/integration_tests /home/jenkins/agent/workspace/jenkins-rapids_premerge-github-2855-ci-2/integration_tests/src/main/python -n 5 -v -rfExXs '-m not premerge_ci_1' --std_input_path=/home/jenkins/agent/workspace/jenkins-rapids_premerge-github-2855-ci-2/integration_tests/src/test/resources --color=yes --test_type pre-commit '-k struct_test or time_window_test'
[2021-09-28T16:23:20.060Z] ImportError while loading conftest '/home/jenkins/agent/workspace/jenkins-rapids_premerge-github-2855-ci-2/integration_tests/src/main/python/conftest.py'.
[2021-09-28T16:23:20.060Z] ../../src/main/python/conftest.py:17: in <module>
[2021-09-28T16:23:20.060Z]     from spark_init_internal import get_spark_i_know_what_i_am_doing
[2021-09-28T16:23:20.060Z] ../../src/main/python/spark_init_internal.py:21: in <module>
[2021-09-28T16:23:20.060Z]     findspark.init()
[2021-09-28T16:23:20.060Z] /usr/local/lib/python3.8/dist-packages/findspark.py:145: in init
[2021-09-28T16:23:20.060Z]     raise Exception(
[2021-09-28T16:23:20.060Z] E   Exception: Unable to find py4j, your SPARK_HOME may not be configured correctly
script returned exit code 4
```

**Comment by gerashegalov** [comment]
build

File: dist/scripts/binary-dedupe.sh [hunk 1]
```diff
@@ -14,62 +14,47 @@
...
+# output is complete, and that the complete output would not exhibit diffs.
+# We compute and compare checksum signatures of same-named classes
 
-# it's now safe to delete duplicate .class files from the original locations
-# don't use rm */% because globbing is broken for files containing $
-for shimDir in $SHIM_DIRS; do
-  xargs --arg-file="$SPARK3XX_COMMON_TXT" -P 4 -n 100 -I% rm "$PARALLEL_WORLDS_DIR/$shimDir/%"
-done
+# The following pipeline determines identical classes across shims in this build.
+# If the files are absent in some shims but are identical in the same shim
```
Comment (tgravescs) [inline dist/scripts/binary-dedupe.sh:36]: sentence seem incomplete?  if they are absent in some but identical in the same then what?

File: dist/scripts/binary-dedupe.sh [hunk 2]
```diff
@@ -14,62 +14,47 @@
...
+#
+echo "Retrieving class files hashing to a single value"
+find . -path './parallel-world/spark*' -type f -name '*class' | \
+  xargs -L 1000 sha1sum -b | \
```
Comment (tgravescs) [inline dist/scripts/binary-dedupe.sh:40]: why the -L 1000?

**Review by tgravescs — APPROVED** [review]

File: dist/scripts/binary-dedupe.sh (hunk 2 above)
Comment (gerashegalov) [reply dist/scripts/binary-dedupe.sh:40]: 1000 is arbitrary and may still blow up with long paths, but this is to limit the number of args per invocation to limit the command line length generated by xargs.

Comment (jlowe) [reply dist/scripts/binary-dedupe.sh:44]: Doesn't xargs already inherently know what the command-line length limit is and will split the input up into batches accordingly?  From the manpage:
```
The command line for command is built up until it reaches a system-defined limit (unless the -n and -L options are used).
```
If our choice is arbitrary, why not leave it to xargs to make it less arbitrary?

Comment (gerashegalov) [reply dist/scripts/binary-dedupe.sh:44]: I am happy to revert this change if it works. I don't recall the particular situation but I definitely have seen xargs fail for me with maximum length exceeded while working on this patch

Comment (jlowe) [reply dist/scripts/binary-dedupe.sh:44]: I'm fine with leaving it in if there's buggy xargs implementations out there.

**Comment by gerashegalov** [comment]
build

**Comment by sameerz** [comment]
Is this required for 21.10?  We should limit 21.10 changes to must have at this point.

**Comment by gerashegalov** [comment]
> Is this required for 21.10? We should limit 21.10 changes to must have at this point.

The first bullet fix in the description is a must for the release, the  rest I can see moving if needed.

**Comment by sameerz** [comment]
> The first bullet fix in the description is a must for the release, the rest I can see moving if needed. But it would be better if we did not split this patch

I am ok not splitting the patch given the size and nature of the change, and keeping in 21.10 given the fix.

**Review by tgravescs — APPROVED** [review]

**Pull request merged**